target/riscv/cpu_helper.c

```diff
@@ -905,6 +966,32 @@ static int riscv_cpu_tlb_fill_impl(CPURISCVState *env, vaddr address, int size,
                     TARGET_FMT_plx " prot %d\n",
                     __func__, im_address, ret, *pa, prot2);
 
+#if defined(TARGET_CHERI) && !defined(TARGET_RISCV32)
+            /*
+             * CHERI's load-side caveats are enforced only on the guest
+             * tables at the moment.  Because we are about to AND the two
+             * prot words together, *set* both caveats in prot2 so that
+             * either bit will be preserved from prot.
+             */
+            prot2 |= PAGE_LC_TRAP | PAGE_LC_CLEAR;
+
+            /*
+             * XXX Eventually we probably want to permit the hypervisor to be
+             * able to force tag clearing or trapping.  In decreasing priority
+             * order, if the tag comes back set...
+             *
+             *  - If the guest says trap, raise a supervisor (VS/S) fault.
+             *
+             *  - If the guest translation succeeds but the hypervisor says
+             *    trap, raise a hypervisor (HS) fault
+             *
+             *  - If either layer says clear, clear the tag.
```

Comment:
Thinking through more, I wonder what should happen if one says clear and the other says trap.. intuitively I'd say if the host says clear but the guest says trap that shouldn't trap, but if the host says trap and the guest says clear it should trap? Hm, which maybe suggests your view of the world was right, as the loaded data flows in the reverse direction to the order you translate (and the order you store). I wonder what Morello does... regardless, it's not specified, so this doesn't need to be completely thought through yet, just an outline of something seemingly sensible.

Reply:
I dropped the "priority order" verbiage and just gave a table of something that seems sensible.  I think "guest says clear host says trap" probably doesn't result in a trap: the guest specifically avoided loading a capability, after all.

Reply:
I think it depends what order you have guest and host traps in; if host traps take precedence, then I think conceptually there has been a tagged capability loaded through the host's PTE and so you should trap, even if the tag is then squashed further up. But leaving precedence for later sounds good to me.